src/main/java/dev/morling/onebrc/CalculateAverage_merykittyunsafe.java

```diff
@@ -0,0 +1,379 @@
+        // 2-byte min + 2-byte max + 8-byte sum + 8-byte count
+        private static final int ENTRY_SIZE = 128;
+        private static final int SIZE_OFFSET = 0;
+        private static final int MIN_OFFSET = 4;
+        private static final int MAX_OFFSET = 6;
+        private static final int SUM_OFFSET = 8;
+        private static final int COUNT_OFFSET = 16;
+        private static final int KEY_OFFSET = 24;
+
+        // There is an assumption that map size <= 10000;
+        private static final int CAPACITY = 1 << 17;
+        private static final int ENTRY_MASK = ENTRY_SIZE * CAPACITY - 1;
+
+        final byte[] data;
+
+        PoorManMap() {
+            this.data = new byte[CAPACITY * ENTRY_SIZE];
+        }
+
+        void observe(long entryOffset, long value) {
+            long baseOffset = Unsafe.ARRAY_BYTE_BASE_OFFSET + entryOffset;
+            UNSAFE.putShort(this.data, baseOffset + MIN_OFFSET,
+                    (short) Math.min(value, UNSAFE.getShort(this.data, baseOffset + MIN_OFFSET)));
+            UNSAFE.putShort(this.data, baseOffset + MAX_OFFSET,
+                    (short) Math.max(value, UNSAFE.getShort(this.data, baseOffset + MAX_OFFSET)));
+            UNSAFE.putLong(this.data, baseOffset + SUM_OFFSET,
+                    value + UNSAFE.getLong(this.data, baseOffset + SUM_OFFSET));
+            UNSAFE.putLong(this.data, baseOffset + COUNT_OFFSET,
+                    1 + UNSAFE.getLong(this.data, baseOffset + COUNT_OFFSET));
+        }
+
+        long indexSimple(long address, int size) {
+            int x;
+            int y;
+            if (size >= Integer.BYTES) {
+                x = UNSAFE.getInt(address);
+                y = UNSAFE.getInt(address + size - Integer.BYTES);
+            }
+            else {
+                x = UNSAFE.getByte(address);
+                y = UNSAFE.getByte(address + size - Byte.BYTES);
+            }
+            int hash = hash(x, y);
+            long entryOffset = (hash * ENTRY_SIZE) & ENTRY_MASK;
+            for (;; entryOffset = (entryOffset + ENTRY_SIZE) & ENTRY_MASK) {
+                int nodeSize = UNSAFE.getInt(this.data, Unsafe.ARRAY_BYTE_BASE_OFFSET + entryOffset + SIZE_OFFSET);
+                if (nodeSize == 0) {
+                    insertInto(entryOffset, address, size);
+                    return entryOffset;
+                }
+                else if (keyEqualScalar(entryOffset, address, size)) {
+                    return entryOffset;
+                }
+            }
+        }
+
+        void insertInto(long entryOffset, long address, int size) {
+            long baseOffset = Unsafe.ARRAY_BYTE_BASE_OFFSET + entryOffset;
+            UNSAFE.putInt(this.data, baseOffset + SIZE_OFFSET, size);
+            UNSAFE.putShort(this.data, baseOffset + MIN_OFFSET, Short.MAX_VALUE);
+            UNSAFE.putShort(this.data, baseOffset + MAX_OFFSET, Short.MIN_VALUE);
+            try (var arena = Arena.ofConfined()) {
```

Comment:
This trick presented by Maurizio Cimadamore depends on the compiler's ability to scalar replace the arena and elide the lifetime check, it will fail miserably if the arena creation is not inlined, though.